com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/graph/ResourceGraph.java

```diff
@@ -0,0 +1,288 @@
+     * to the graph.
+     * @param rootResource The resource to create graph from.
+     */
+    public void add(IResource rootResource) throws CompileExceptionError {
+        ResourceWalker.walk(project, rootResource, this);
+    }
+
+    // used in tests
+    public ResourceNode add(String resourcePath, ResourceNode parentNode) {
+        IResource currentResource = project.getResource(resourcePath);
+        IResource parentResource = project.getResource(parentNode.getPath());
+        ResourceNode currentNode = resourceToNodeLookup.get(currentResource);
+        if (currentNode == null) {
+            currentNode = new ResourceNode(resourcePath);
+            resourceToNodeLookup.put(currentResource, currentNode);
+            pathToNodeLookup.put("/" + currentResource.getPath(), currentNode);
+            resourceNodes.add(currentNode);
+            if (currentResource.getPath().endsWith("collectionproxyc")) {
+                currentNode.setType(ResourceNode.Type.CollectionProxy);
+            }
+        }
+        parentNode.addChild(currentNode);
+        return currentNode;
+    }
+    public ResourceNode add(ResourceNode resourceNode, ResourceNode parentNode) {
+        return add(resourceNode.getPath(), parentNode);
+    }
+
+    /**
+     * Get the root resource node of the graph. All resources added to the graph
+     * will exist as children of the root resource.
+     * @return The root node
+     */
+    public ResourceNode getRootNode() {
+        return root;
+    }
+
+    /**
+     * Get a collection of all resources added to the graph.
+     * @return A collection of resources
+     */
+    public Collection<IResource> getResources() {
+        return resources;
+    }
+
+    /**
+     * Get resource node from path
+     * @param path The path to get resource node for
+     * @return The resource node for the path
+     */
+    public ResourceNode getResourceNodeFromPath(String path) {
+        return pathToNodeLookup.get(path);
+    }
+
+    /**
+     * Set hex digests for all resource nodes in the graph
+     * @param hexDigests Map with hex digests, keyed on relative resource paths
+     */
+    public void setHexDigests(Map<String, String> hexDigests) {
+        for (ResourceNode resourceNode : resourceNodes) {
+            String hexDigest = hexDigests.get(resourceNode.getPath());
+            resourceNode.setHexDigest(hexDigest);
+        }
+    }
+
+    private void flagAllNonExcludedResources(ResourceNode node, HashSet<ResourceNode> visited) {
+        for (ResourceNode child : node.getChildren()) {
+            if (visited.contains(child)) {
+                continue;
+            }
+            child.flagAsUsedInMainBundle();
+            visited.add(child);
+            if (child.checkType(ResourceNode.Type.CollectionProxy) && child.getExcludedFlag()) {
+                continue;
+            }
+            flagAllNonExcludedResources(child, visited);
+        }
+    }
```

Comment:
Visit all the bodes in non-excluded collections once and flag them as used in the main bundle. We can exclude everything else.